pkg/controller/bootstrap/tokencleaner.go

```diff
@@ -77,37 +81,57 @@ func NewTokenCleaner(cl clientset.Interface, options TokenCleanerOptions) (*Toke
 		}
 	}
 
-	secretSelector := fields.SelectorFromSet(map[string]string{api.SecretTypeField: string(bootstrapapi.SecretTypeBootstrapToken)})
-	e.secrets, e.secretsController = cache.NewInformer(
-		&cache.ListWatch{
-			ListFunc: func(lo metav1.ListOptions) (runtime.Object, error) {
-				lo.FieldSelector = secretSelector.String()
-				return e.client.CoreV1().Secrets(e.tokenSecretNamespace).List(lo)
+	secrets.Informer().AddEventHandlerWithResyncPeriod(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1.Secret:
+					return t.Type == bootstrapapi.SecretTypeBootstrapToken && t.Namespace == e.tokenSecretNamespace
+				default:
+					utilruntime.HandleError(fmt.Errorf("object passed to %T that is not expected: %T", e, obj))
+					return false
+				}
 			},
-			WatchFunc: func(lo metav1.ListOptions) (watch.Interface, error) {
-				lo.FieldSelector = secretSelector.String()
-				return e.client.CoreV1().Secrets(e.tokenSecretNamespace).Watch(lo)
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc:    e.evalSecret,
```

Comment:
I don't like it. evelSecret is non-trivial function (in particular it may send API call). And note that handler functions are invoked synchronously from informers so long processing here will result is delaying all other handlers registered for secrets.

This should be changed to:
1. AddFunc should just store the secret in some queue or sth like that
2. There should be a separate goroutine(s) that will be picking things from that queue and calling evalSecret for them.

Reply:
👍, let's use a work queue like many of the other controllers do